src/customers/controller/customer.controller.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  ApiTags,
+  ApiOperation,
+  ApiResponse,
+  ApiParam,
+  ApiBody,
+} from '@nestjs/swagger';
+import {
+  Body,
+  Controller,
+  Delete,
+  Get,
+  HttpCode,
+  Param,
+  Patch,
+  Post,
+} from '@nestjs/common';
+
+import { CustomerService } from '../service/customer.service';
+import { CustomerMapper } from '../mapper/customer.mapper';
+import { CustomerDto } from '../dto/customer.dto';
+import { Customer } from '../domain/customer.entity';
+import { CreateCustomerDto } from '../dto/create-customer.dto';
+import { UpdateCustomerDto } from '../dto/update-customer.dto';
+
+@ApiTags('customers')
+@Controller('customers')
+export class CustomerController {
+  constructor(private readonly customerService: CustomerService) {}
+
+  @Get(':customerId')
+  @HttpCode(200)
+  @ApiOperation({ summary: 'Get customer by ID' })
+  @ApiParam({
+    name: 'customerId',
+    description: 'The unique identifier of the customer',
+  })
+  @ApiResponse({
+    status: 200,
+    description: 'Customer found',
+    type: CustomerDto,
+  })
+  @ApiResponse({ status: 404, description: 'Customer not found' })
+  async getById(@Param('customerId') customerId: string): Promise<CustomerDto> {
+    return CustomerMapper.toDto(await this.customerService.getById(customerId));
+  }
+
+  @Get()
+  @HttpCode(200)
+  @ApiOperation({ summary: 'Get all customers' })
+  @ApiResponse({
+    status: 200,
+    description: 'List of customers',
+    type: [CustomerDto],
+  })
+  async getAll(): Promise<CustomerDto[]> {
+    const customers = await this.customerService.getAll();
+    return customers.map((customer: Customer) =>
+      CustomerMapper.toDto(customer),
+    );
+  }
+
+  @Post('register')
+  @HttpCode(201)
+  @ApiOperation({ summary: 'Register a new customer' })
+  @ApiBody({ type: CreateCustomerDto })
+  @ApiResponse({ status: 201, description: 'Customer registered successfully' })
+  @ApiResponse({ status: 400, description: 'Invalid input data' })
+  async register(
+    @Body() registrationDto: CreateCustomerDto,
+  ): Promise<{ message: string }> {
+    await this.customerService.create(
+      CustomerMapper.fromCreateDto(registrationDto),
+    );
+
+    return {
```

Comment:
**issue:** note that usually you want to return the successfully created object in your request to be later used by the client (e.g.: it will likely need to newly created/updated object to update something in the UI).

Note: for customer you will need to take care you don't expose the hashed password by mistake

Reply:
Applies to some other places in your controllers where you have done this

Reply:
Thanks for the remark! In this particular case I thought that, since the newly created customer would need to login after a registration, there would be no need to return the customer object. But you are right, as our auth service, for example, will need some info about the newly created customer to build up the token payload ( eg. id for the token subject)

Reply:
Also, regarding the updated entities, I have initially thought of returning them but then removed the return of the updated entity, since I considered it redundant, as for some frontend querying solutions we might invalidate the cached entity and then submit another, separate, query to refetch the updated entity. But its really helpful to see that returning the updated entity is recommended. After all, the backend should be, to some extent, independent of the frontend/consumer solution, as that might change in the future